code/Serving_API.py

```diff
@@ -1,27 +1,258 @@
+import sqlite3
+import networkx as nx
+import csv
+from xml.dom.minidom import Document
+from haversine import haversine
+
+# find the closest point for a given coordinate in case it is not in the graph
+
+
+def find_closest_point(point, points_set):
+    return min(points_set, key=lambda p: haversine(point, p))
+
+# helper function to build nx graph with src/dst city, src/dst coordinates, wkt path, cabel type and distance.
+
+
+def add_edge(G, city1, city2, distance, path_wkt, src_city_coord, dst_city_coord, cable_type):
+    # Add or update nodes with their coordinates
+    G.add_node(city1, coord=src_city_coord)
+    G.add_node(city2, coord=dst_city_coord)
+
+    # Add the edge with its properties
+    G.add_edge(city1, city2, weight=distance, path_wkt=path_wkt,
+               src_city_coord=src_city_coord, dst_city_coord=dst_city_coord, cable_type=cable_type)
+
+# read standard paths from database
+
+
+def get_all_standard_paths(db_file):
+    conn = sqlite3.connect(db_file)
+    cursor = conn.cursor()
+    sql_query = """
+    SELECT sp.from_city, sp.from_state, sp.from_country, sp.to_city, sp.to_state, sp.to_country, sp.distance_km, sp.path_wkt
+    FROM standard_paths sp
+    """
+    cursor.execute(sql_query)
+    datas = cursor.fetchall()
+    conn.close()
+    return datas
+
+# lati and longti in wkt path is reversed, use this helper function to reverse that
+
+
+def coordinate_reverser(coord):
+    coord_longti, coord_lati = coord
+    return (coord_lati, coord_longti)
+
+# reverse linestring coords order for direct graph
+
+
+def reverse_linestring_order(linestring):
+    # Remove the 'LINESTRING(' prefix and the closing ')'
+    clean_string = linestring.replace("LINESTRING(", "").replace(")", "")
+
+    # Split the string into coordinate pairs
+    coords = clean_string.split(", ")
+
+    # Reverse the order of the coordinate pairs
+    reversed_coords = coords[::-1]
+
+    # Join the reversed coordinates back into a LineString
+    reversed_linestring = "LINESTRING(" + ", ".join(reversed_coords) + ")"
+
+    return reversed_linestring
+
+# helper function to get the src/dst coordinate from a wkt path
+
+
+def parse_wkt_linestring(wkt_string):
+    # Remove the LINESTRING prefix and split the string into coordinate pairs
+    coord_pairs = wkt_string.replace(
+        "LINESTRING(", "").replace(")", "").split(", ")
+
+    # Extract the first and last coordinate pairs
+    first_pair = coord_pairs[0]
+    last_pair = coord_pairs[-1]
+
+    # Convert the coordinate pairs from strings to tuples of floats
+    start_city_coord = tuple(map(float, first_pair.split(" ")))
+    end_city_coord = tuple(map(float, last_pair.split(" ")))
+
+    return start_city_coord, end_city_coord
+
+
+def graph_build_helper(G, coordCityMap, coord_set, path_set, option):
+    for from_city, from_state, from_country, to_city, to_state, to_country, distance_km, path_wkt in path_set:
+
+        start_city_coord, end_city_coord = parse_wkt_linestring(path_wkt)
```

Comment:
Your parse_wkt_linestring in fact does not exact do the name suggest.

Since you need this info later for reverse, I'd suggest that this function returns a `ls: shapely.geometry.LineString` object, and you can directly get `start_city_coord = coordinate_reverse(ls.coords[0])` and `end_city_coord = coordinate_reverse(ls.coords[-1])`, and at the end, oin the `add_edge` call, use `ls.wkt` and `LineString([ls.coords[::-1]).wkt`.

Reply:
Actually, I think it's better to store the LineString object in the graph, so that later, when creating the `wkt_list`, we can use standard type `MultiLineString`. I'll add a comment there.